File: src/main/python/problem3/largest_prime_factor.py
# Claire Jaja
# 11/1/2014
#
# Project Euler
# Problem 3
# Largest prime factor
#
# The prime factors of 13195 are 5, 7, 13, and 29.
# What is the largest prime factor of the number 6008514745143?

import math
from collections import defaultdict

def main():
    #num = 13195
    num = 600851475143
    primes = calc_primes(int(math.sqrt(num))+1)
    primes.sort()
    primes.reverse()
    largest_prime_factor = 0
    i = 0
    while not largest_prime_factor:
        if float(num) % primes[i] == 0:
            largest_prime_factor = primes[i]
        else:
            i += 1
    print(largest_prime_factor)

# calculate all primes up to x using sieve of Eratosthenes
def calc_primes(x):
    is_prime = defaultdict(lambda:True)
    for i in range(3,int(math.sqrt(x))+1,2):
        if is_prime[i]:
            for j in range(i,int(x/i)+1,2):
                is_prime[i*j] = False
    primes = [y for y in range(1,x+1) if y % 2 != 0 and is_prime[y]]
    primes.append(2)
    return primes

if __name__ == "__main__":
    main()

File: src/main/scala/problem3/LargestPrimeFactor.scala
/**
* @author Claire Jaja
* @version 10/30/14
*
* Project Euler
* Problem 3
* Largest prime factor
*
* The prime factors of 13195 are 5, 7, 13, and 29.
* What is the largest prime factor of the number 600851475143?
*/
import math._
import scala.collection.mutable

object LargestPrimeFactor {
  val num = 600851475143L
  //val num = 13195

  def main(args: Array[String]) {
    // calculate all primes up to square root of num
    var primes = calcPrimes(sqrt(num.toDouble).ceil.toInt)
    var largestPrimeFactor = 0
    var i = primes.size - 1

    for (prime <- primes) {
      if (num % prime == 0 && prime > largestPrimeFactor) largestPrimeFactor = prime
    }

    println(largestPrimeFactor)
  }

  // calculate all primes up to x using sieve of Eratosthenes
  def calcPrimes(x: Int): List[Int] = {
    var isPrime = mutable.Map((for (i <- 3 to x by 2) yield (i -> true)):_*)
    for (i <- 3 to sqrt(x.toDouble).ceil.toInt by 2) {
      if (isPrime(i)) {
        for (j <- i to x/i by 2) {
          isPrime(i*j) = false
        }
      }
    }
    
    return isPrime.filter({case (num,bool) => bool}).map({case (num,bool) => num}).toList
  }
}
